src/cleanfigure.m

```diff
@@ -367,6 +368,259 @@ function coarsenLine(meta, handle, minimumPointsDistance)
+
+  % Set the new (masked) data.
+  set(handle, 'XData', xData);
+  set(handle, 'YData', yData);
+
+  return;
+end
+% =========================================================================
+function a = featureArea(x,y)
+  % Performs Visvalingam's algorithm:
+  % Given the path defined by x,y, the function returns list of the
+  % maximum area associated with each point in the list. Path then can be
+  % simplified via x = x(a>tol)
+  %
+  % Runtime is O(n log(n))
+  %
+  % Used by 'simplifyLine'. 
+
+  n = numel(x);
+  
+  % Index of next and previous elements in the linked list which defines
+  % the path
+  llst = [(0:n-1)',[(2:n)';0]];
+  
+  % 'heap' stores the points in an array heap, referenced by their index
+  % First and last points are assumed fixed and not added to the heap
+  % 'pos' stores the current position of each point in the heap array
+  % will be needed to lookup position of updated points
+  heap = (2:n-1)';
+  len = numel(heap);
+  pos = (1:len)';
+
+
+  area = @(i,j,k) abs((y(j) - y(k)).*x(i) + (y(k)-y(i)).*x(j) + ...
+                                                    (y(i)-y(j)).*x(k))/2;
+
+  % Endpoints are assigned infinte area so they can't be removed
+  a = [Inf;area((1:n-2)', (2:n-1)', (3:n)')';Inf];
+
+  % Keep track of the maximum area removed so far to ensure 
+  % a given element will only be excluded after earlier elements
+  maxArea = 0;
+
+  %Heapify the 'heap' array based on the area, using Floyd's alg
+  root = bitshift(len,-1); %starting with the first parent node
+  while root >= 1
+    down(root);
+    root = root-1;
+  end
+  
+
+  %Now remove the element with the minimum area off the heap
+  while len > 1
+     if a(heap(1)) > maxArea
+       maxArea = a(heap(1));
+     else
+       % ensure current element will only be excluded when earlier elements
+       % are also excluded
+       a(heap(1)) = maxArea;
+     end
+
+    % remove smallest element from heap
+    e = pop(1);
+
+   
+    left = llst(e,1);
+    right = llst(e,2);
+    % remove element from linked list
+    llst(left,2) = llst(e,2);
+    llst(right,1) = llst(e,1);
+
+
+    %Update area of neighbouring points if they're not the ends points
+    if llst(left,1) > 0
+      a(left) = area(llst(left,1),left,llst(left,2));
+      pop(pos(left-1));
+      push();
+    end
+
+    if llst(right,2) > 0
+      a(right) = area(llst(right,1),right,llst(right,2));
+      pop(pos(right-1));
+      push();
+    end
+  end
+
+  if numel(heap) >0 &&  a(heap(1)) < maxArea
+    a(heap(1)) = maxArea;
+  end
+
+  % Move element at "root" down the heap, assuming the heap property
+  % is satisfied for the rest of the tree
+  function down(root)
+    while 2*root  <= len %while the root has a child
+      lchild = 2*root; 
+      rchild = lchild+1;
+      % Find the minimum of the root and its children
+      swap = root;
+      if a(heap(lchild)) < a(heap(swap)) ...
+          || (a(heap(lchild)) == a(heap(swap)) && rand(1)>0.5)
```

Comment:
Same remark as https://github.com/matlab2tikz/matlab2tikz/pull/571/files#r26300183: `rand` is bad for testing.